sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+    }
+
+    public Read withTopic(String topic) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    public Read withMaxNumRecords(long maxNumRecords) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    public Read withMaxReadTime(Duration maxReadTime) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
+    }
+
+    @Override
+    public PCollection<JmsRecord> apply(PBegin input) {
+      // handles unbounded source to bounded conversion if maxNumRecords is set.
+      Unbounded<JmsRecord> unbounded = org.apache.beam.sdk.io.Read.from(createSource());
+
+      PTransform<PInput, PCollection<JmsRecord>> transform = unbounded;
+
+      if (maxNumRecords < Long.MAX_VALUE) {
+        transform = unbounded.withMaxNumRecords(maxNumRecords);
+      } else if (maxReadTime != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime);
+      }
+
+      return input.getPipeline().apply(transform);
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      if (connectionFactory == null) {
+        throw new IllegalArgumentException("ConnectionFactory not specified");
+      }
+      if (queue == null && topic == null) {
+        throw new IllegalArgumentException("Either queue or topic not specified");
+      }
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+
+      if (queue != null) {
+        builder.add(DisplayData.item("queue", queue));
+      }
+      if (topic != null) {
+        builder.add(DisplayData.item("topic", topic));
+      }
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////
+
+    private ConnectionFactory connectionFactory;
+    private String queue;
+    private String topic;
+    private long maxNumRecords;
+    private Duration maxReadTime;
+
+    private Read() {}
+
+    private Read(
+        ConnectionFactory connectionFactory,
+        String queue,
+        String topic,
+        long maxNumRecords,
+        Duration maxReadTime) {
+      super("JmsIO.Read");
+
+      this.connectionFactory = connectionFactory;
+      this.queue = queue;
+      this.topic = topic;
+      this.maxNumRecords = maxNumRecords;
+      this.maxReadTime = maxReadTime;
+    }
+
+    /**
+     * Creates an {@link UnboundedSource<JmsRecord, ?>} with the configuration in
+     * {@link Read}. Primary use case is unit tests, should not be used in an
+     * application.
+     */
+    @VisibleForTesting
+    UnboundedSource<JmsRecord, JmsCheckpointMark> createSource() {
+      return new UnboundedJmsSource(
+          connectionFactory,
+          queue,
+          topic);
+    }
+
+  }
+
+  private JmsIO() {}
+
+  private static class UnboundedJmsSource extends UnboundedSource<JmsRecord, JmsCheckpointMark> {
+
+    private final ConnectionFactory connectionFactory;
```

Comment:
Finally, I double checked and according to the javadoc, JMS providers have to provide serializable `ConnectionFactory`.